flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java

```diff
@@ -67,13 +77,33 @@ class Executing extends StateWithExecutionGraph implements ResourceListener {
         this.context = context;
         Preconditions.checkState(
                 executionGraph.getState() == JobStatus.RUNNING, "Assuming running execution graph");
+        this.scalingIntervalMin = scalingIntervalMin;
+        this.scalingIntervalMax = scalingIntervalMax;
+        Preconditions.checkState(
+                !scalingIntervalMin.isNegative(),
+                "{} must be positive integer or 0",
+                JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MIN.key());
+        if (scalingIntervalMax != null) {
+            Preconditions.checkState(
+                    scalingIntervalMax.compareTo(scalingIntervalMin) > 0,
+                    "{}({}) must be greater than {}({})",
+                    JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MAX.key(),
+                    scalingIntervalMax,
+                    JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MIN.key(),
+                    scalingIntervalMin);
+        }
 
         deploy();
 
         // check if new resources have come available in the meantime
         context.runIfState(this, this::maybeRescale, Duration.ZERO);
```

Comment:
After the `Executing` is created, it means the it's a new rescale. We can rescale after `scalingIntervalMin` even if the new resources have come, right? If yes, we should call `rescaleWhenCooldownPeriodIsOver()` here.

Reply:
not necessary a new rescale it could be a transition from Created to Executing state (FLIP-160). In that case we want to eagerly adapt to resources that could have arrived when while we were starting and not wait for the cooldown period to end.

Reply:
Sorry, I don't understand why we should call `maybeRescale` instead of  `rescaleWhenCooldownPeriodIsOver`.

As I mentioned before: `After the Executing is created, it means the it's a new rescale. We can rescale after scalingIntervalMin even if the new resources have come, right?`

If no, the scalingIntervalMin cannot be respected.

Please correct me if I'm wrong.

Reply:
Any progress?

Reply:
when the Executing is initially created (without any restart) meaning when the job transitions from Creating to Executing, what is the point in waiting the cooldown period ? In the case of a restart, the scalingIntervalMin is respected because the triggering of the restart (maybeRescale) was not called before scalingIntervalMin during the life of the pipeline. Then there is no point in waiting again when we recreate the Executing object (we have already waited before the restart).

Reply:
When the `Executing` is created, it means this is a new restart, we still need to respect the `scalingIntervalMin`.

If we call `maybeRescale` here, AdaptiveScheduler may not guarantee scalingIntervalMin. Following is an example:

- 2 slots comes when job transitions from Creating to Executing.
- The job will go rescale directly if `maybeRescale` is called, because 2 > `min-parallelism-increase`.
- So this round, the Executing ignore the `scalingIntervalMin`.
- This can happen all the time when 2 new slots arrive and the job transitions from Creating to Executing.

IIUC, when the `Executing` is created, it means this is a new restart, we should ensure it can run at least `scalingIntervalMin`.

Please let me know if my understading is wrong or it's not clear.

Reply:
done